intellij-toml/src/main/kotlin/org/toml/lang/CrateVersion.kt

```diff
@@ -0,0 +1,161 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.toml.lang
+
+// Based on https://github.com/steveklabnik/semver/blob/master/src/version.rs
+sealed class Identifier : Comparable<Identifier> {
+    companion object {
+        fun parse(input: String): Identifier {
+            // Strings such as 0851523 should be parsed as AlphaNumeric because Numeric would lose the 0 in front
+            val int = if (input.startsWith("0")) null else input.toIntOrNull()
+            return if (int != null) {
+                Numeric(int)
+            } else {
+                AlphaNumeric(input)
+            }
+        }
+    }
+
+    data class Numeric(val num: Int) : Identifier() {
+        override fun compareTo(other: Identifier): Int {
+            if (other is Numeric) {
+                return this.num.compareTo(other.num)
+            }
+
+            return 0
+        }
+
+        override fun toString() = "$num"
+    }
+
+    data class AlphaNumeric(val string: String) : Identifier() {
+        init {
+            require(string.isNotEmpty()) { "Identifier can't be empty" }
+        }
+
+        override fun compareTo(other: Identifier): Int {
+            if (other is AlphaNumeric) {
+                return this.string.compareTo(other.string)
+            }
+
+            return 0
+        }
+
+        override fun toString() = string
+    }
+}
+
+data class IncompleteCrateVersion(val major: Int?, val minor: Int?, val patch: Int?, val pre: List<Identifier> = emptyList(), val build: List<Identifier> = emptyList()) {
```

Comment:
This class looks weird. Maybe use one class (`CrateVersion`) for all purposes? I mean something like this:
```Kotlin
data class CrateVersion(
    val major: Int,
    val minor: Int,
    val patch: Int,
    val pre: List<Identifier> = emptyList(),
    val build: List<Identifier> = emptyList()
) : Comparable<CrateVersion> {
    data class PartialCrateVersion(
        val major: Int?,
        val minor: Int?,
        val patch: Int?,
        val pre: List<Identifier> = emptyList(),
        val build: List<Identifier> = emptyList()
    ) {
        fun tryConvertToCrateVersion(): CrateVersion? {
            return if (major != null && minor != null && patch != null) {
                CrateVersion(major, minor, patch, pre, build)
            } else {
                null
            }
        }
    }

    companion object {
        fun parse(string: String): CrateVersion {
            val partialVersion = parsePartial(string)
            return tryConvertToCrateVersion(partialVersion) ?: throw IllegalArgumentException("Incomplete version")
        }

        fun parsePartial(string: String): PartialCrateVersion {
            // Your IncompleteCrateVersion.parse
        }
```

Reply:
BTW, have you tried `com.intellij.util.text.SemVer` for your purposes? It is widely used in IntelliJ Rust plugin.